File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};
    static int R;
    static int C;

    public int numIslands(char[][] grid) {
        R = grid.length;
        C = grid[0].length;
        int ans = 0;
        for (int i = 0; i < R; i++){
            for (int j = 0; j < C; j++){
                if (grid[i][j] == '1'){
                    dfs(grid, i, j);
                    ans++;
                }
            }
        }
        return ans;
    }
    public void dfs(char[][] grid, int r, int c) {
        grid[r][c] = '0';
        for (int i = 0; i < 4; i++){
            int newR = r + dr[i];
            int newC = c + dc[i];
            if (0 <= newR && newR < R && 0 <= newC && newC < C && grid[newR][newC] == '1'){
                dfs(grid, newR, newC);
            }
        }
    }
}

File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def dfs(row, col):
            for i in range(4):
                next_row = dx[i] + row
                next_col = dy[i] + col

                if (
                    next_row < 0
                    or next_row >= m
                    or next_col < 0
                    or next_col >= n
                    or grid[next_row][next_col] == "0"
                ):
                    continue

                grid[next_row][next_col] = "0"
                dfs(next_row, next_col)

        n = len(grid[0])  # col
        m = len(grid)  # row
        answer = 0
        # visited = [[False] * n for _ in range(m)]
        # 방향 벡터
        dx = [-1, 1, 0, 0]
        dy = [0, 0, -1, 1]

        for i in range(m):
            for j in range(n):
                if grid[i][j] == "1":
                    answer += 1
                    dfs(i, j)
                    
        return answer